engines/docker/image_cache.go

```diff
@@ -0,0 +1,108 @@
+package dockerengine
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+	"github.com/taskcluster/taskcluster-worker/runtime/atomics"
+	"github.com/taskcluster/taskcluster-worker/runtime/caching"
+	"github.com/taskcluster/taskcluster-worker/runtime/gc"
+)
+
+const dockerPullImageInactivityTimeout = 30 * time.Second
+
+type imageResource struct {
+	Repository string
+	Tag        string
+	Size       uint64
+	client     *docker.Client
+	monitor    runtime.Monitor
+	dispose    atomics.Once
+}
+
+type cacheContext struct {
+	context.Context
+}
+
+func newCacheContext(ctx context.Context) *cacheContext {
+	if ctx != nil {
+		return &cacheContext{
+			Context: ctx,
+		}
+	}
+	return nil
+}
+
+func (ctx *cacheContext) Progress(description string, percent float64) {
+	debug("progress: %s %f", description, percent)
+}
+
+func (im *imageResource) MemorySize() (uint64, error) {
+	return im.Size, nil
+}
+
+func (im *imageResource) DiskSize() (uint64, error) {
+	return 0, gc.ErrDisposableSizeNotSupported
+}
+
+func (im *imageResource) Dispose() error {
+	var err error
+	im.dispose.Do(func() {
+		err = im.client.RemoveImage(buildImageName(im.Repository, im.Tag))
```

Comment:
We should probably do some testing on this...

In this case `Tag` is a hash, right?

We could have a scenario where:
`hashA <- hashB <- hashC`

So `hashA` is the base image, `hashB` is built from A and `hashC` is build from B.

cases to consider:
 * If I fetch `hashB` I'll get both `hashA` and `hashB`... but if I delete `hashB`, do I get both removed?
 * If I fetch `hashC` I'll get both `hashA`, `hashB` and `hashC`, if I then fetch `hashB` and remove `hashC` will I remove `hashB` ?

We should probably double check this... and check if docker actually removes all the intermediary image layers...
I think docker tracks which image layers where explicitly pulled and has to be explicitly removed, before the deletes the underlying layers... So I think it works... but we should probably double check :)

Maybe write a comment here on what behavior we have from docker, and why this works... even though we might have two `imageResource`s that uses the same base image layers... and that those base image layers won't disappear until they are both remoted... (ie. that docker does some reference counting for us)

Reply:
`Remove an image, along with any untagged parent images that were referenced by that image.

Images can't be removed if they have descendant images, are being used by a running container or are being used by a build.
`
So, if `hashA`<-`hashB`<-`hashC`, `hashB` cannot be removed since it has a descendant image `hashC`. In the second case, removing `hashC` should remove `hashB` if it has no other descendant images, but if we have fetched `hashB` and are using it in a container, then it won't be removed. Please note that I am not using the Force flag when removing images.

Reply:
Yeah, looking [examples](https://docs.docker.com/engine/reference/commandline/rmi/#examples), it seems docker does the right thing.